src/napari_spatialdata/_view.py

```diff
@@ -116,12 +171,167 @@ def __init__(self, napari_viewer: Viewer, model: DataModel | None = None):
+                else:
+                    # change widget data if the annotation already exists and is selected
+                    if (self.color_widget.widget.getAttribute() == "obs") and (
+                        self.color_widget.widget.chosen == self.annotation_name
+                    ):
+                        if self.color_widget.widget.data is not None:
+                            self.color_widget.widget.data["vec"] = self.selected_vector
+
+                        color_label = self.color_widget.getFormattedLabel()
+                        if color_label is not None:
+                            color_label = color_label + "__change__"
+
+                        # automatically replot - may revisit later for performance
+                        self.plot_widget._onClick(
+                            self.x_widget.widget.data,
+                            self.y_widget.widget.data,
+                            self.color_widget.widget.data,
+                            self.x_widget.getFormattedLabel(),
+                            self.y_widget.getFormattedLabel(),
+                            color_label,
+                        )
+
+                    self.change_status("Annotation updated.")
+
+            # display status message that no column name was provided
+            else:
+
+                self.change_status("No column name provided.")
+
+        # display status message - no rois
+        else:
+
+            self.change_status("No rois selected.")
+
+    def save_sdata(self) -> None:
+        """Save sdata or AnnData."""
+        # data taken from the viewer
+        if self._viewer is not None:
+
+            selected_layer = self._viewer.layers.selection.active
+            selected_table = self.table_name_widget.currentText()
+
+            # trigger saving of the table
+            # TODO: have to find another way to pass this as this is deprecated from 0.5.0 onwards
+            # it's used in the save_sdata method
+            self._viewer_model = self._viewer.window._dock_widgets["SpatialData"].widget().viewer_model
+            self._viewer_model._write_element_to_disk(
+                selected_layer.metadata["sdata"],
+                selected_table,
+                selected_layer.metadata["sdata"][selected_table],
+                overwrite=True,
+            )
+        # data provided directly as AnnData
+        else:
+            d = AnnDataSaveDialog()
+            file_path = d.show_dialog()
+
+            if file_path:
+                if file_path.endswith(".h5ad") or file_path.endswith(".zarr") or file_path.endswith(".csv"):
+                    if file_path.endswith(".h5ad"):
+                        self.model.adata.write(file_path)
+                    if file_path.endswith(".zarr"):
+                        self.model.adata.write_zarr(file_path)
+                    if file_path.endswith(".csv"):
+                        self.model.adata.write_csvs(file_path)
```

Comment:
Let's bring also this comment to the conversation because initiating and saving are related:

> I would only make one thing clearer in the "scatterwidget.ipynb" and "scatterwidget_annotation.ipynb" notebooks.
I would clarify a bit around the scenarios: "napari is running" vs "no napari viewer detected" and the case in which the table is from a standalone AnnData vs coming from a SpatialData object.
In principle there are 4 cases. Are these two possible: "napari is running" and "the table is from a standalone AnnData" and "no napari viewer detected" and "the table is from a SpatialData object?

For simplicity, I suggest that we provide the user with only two options:

1. Running the Scatter Widget with AnnData.
2. Running Interactive mode with sdata.

For option 1, we do not specify the source of the AnnData, as we currently do not support modification of an existing AnnData file. This will hopefully be revisited in the future when such functionality becomes available. AnnData can either come directly from a file or be derived from an sdata structure, but in both cases, it is detached from its origin. It can be saved as a new file, or the user may overwrite the existing AnnData file (after acknowledging the OS's generic warning—AnnDataSaveDialog provides this functionality). If the AnnData is derived from sdata, it is the user's responsibility to reattach the modified table to the original sdata, if desired. If we keep this behavior, the `AnnDataSaveDialog` has a totally different purpose than `SaveDialog`.

If a user wishes to interact with the sdata object, it should be done through Interactive mode and not solely through the Scatter Widget. I don't see any downside to this approach, but I am curious about your opinion.

For option 2, when the Scatter Widget is used as part of the Interactive viewer, it obtains sdata from the layer on which it is called. The status of the Scatter Widget indicates whether the sdata from the layer is backed. When a user clicks 'Save', the sdata will be overwritten if backed; otherwise, a logger will display the message: "Spatialdata object is not stored on disk, can only add element in memory."

On a technical level:

The Annotation button action does not depend on the sdata status. When the button is clicked:
- The AnnData of the Scatter Widget is modified.
- The AnnData of the layer is modified.
- The sdata of the layer is modified (a merge of the table is performed).

The 'Save' button calls `_write_element_to_disk` (similar to the AnnotationWidget), which is a convenient place to add additional precautions, such as saving a temporary table and renaming it later.

Since the Scatter Widget always works with an existing table, overwriting a table in sdata is unavoidable. Given that every click of the 'Save' button would result in an overwrite, should we consider adding a warning to the notebooks and documentation instead of displaying a dialog box? This might be a better approach, as there isn’t much the user can do to make the process safer at this point. What do you think?

Reply:
(the conversation can be resolved)